signac/contrib/job.py

```diff
@@ -131,10 +131,16 @@ def _save(self):
                 raise
 
         # Since all the jobs are equivalent, just grab the filename from the
-        # last one and init it. Also migrate the lock for multithreaded support.
-        old_lock_id = self._lock_id
+        # last one and init it. Also migrate the lock for multithreaded
+        # support. The sequence of operations here is imoprtant since changing
+        # the filename updates the lock id.
+        if type(self)._threading_support_is_active:
+            old_lock_id = self._lock_id
+
         self._filename = job._statepoint_filename
-        type(self)._locks[self._lock_id] = type(self)._locks.pop(old_lock_id)
+
+        if type(self)._threading_support_is_active:
+            type(self)._locks[self._lock_id] = type(self)._locks.pop(old_lock_id)
```

Comment:
This segment feels like implementation details of the base class are leaking here. Maybe this could be abstracted into a classmethod in the base class? That would also make direct class attribute access via `type(...)` unnecessary. What do you think?

Reply:
I agree. I've moved this logic into a property setter for `filename` in `JSONCollection`, which is the most appropriate location since that class is where the `filename`<-->`lock` relationship is established.

Reply:
Great, thanks.